src/BotPlutusInterface/TimeSlot.hs

```diff
@@ -0,0 +1,174 @@
+posixTimeToSlotIO ::
+  PABConfig ->
+  Ledger.POSIXTime ->
+  IO (Either TimeSlotConversionError Ledger.Slot)
+posixTimeToSlotIO pabConf pTime = runEitherT $ do
+  nodeInfo <- liftIO $ mkNodeInfo pabConf
+  eraHist <- newET (queryEraHistory nodeInfo)
+  sysStart <- newET $ querySystemStart nodeInfo
+  firstEitherT toError . hoistEither $
+    posixTimeToSlot sysStart eraHist pTime
+
+{- | Convert a `POSIXTimeRange` to `SlotRange`.
+ Gives the biggest slot range that is entirely contained by the given time range.
+-}
+
+-- Analogous to `posixTimeRangeToContainedSlotRange` from plutus-ledger
+posixTimeRangeToContainedSlotRangeIO ::
+  PABConfig ->
+  Ledger.POSIXTimeRange ->
+  IO (Either TimeSlotConversionError Ledger.SlotRange)
+posixTimeRangeToContainedSlotRangeIO
+  pabConf
+  ptr@(Interval (LowerBound start startIncl) (UpperBound end endIncl)) = runEitherT $ do
+    -- getting required info from node
+    nodeInfo <- liftIO $ mkNodeInfo pabConf
+    sysStart <- newET $ querySystemStart nodeInfo
+    eraHsitory <- newET (queryEraHistory nodeInfo)
+    let epochInfo = toLedgerEpochInfo eraHsitory
+        pparams =
+          CAPI.toLedgerPParams
+            CAPI.ShelleyBasedEraAlonzo
+            (pcProtocolParams pabConf)
+
+    let extTimeToExtSlot = convertExtended sysStart eraHsitory
+        getClosure = getExtClosure pparams epochInfo sysStart
+
+    -- conversions
+    startSlot <- extTimeToExtSlot start
+    startSlotClosure <- getClosure startSlot startIncl
+    endSlot <- extTimeToExtSlot end
+    endSlotClosure <- getClosure endSlot endIncl
+    let lowerB = LowerBound startSlot startSlotClosure
+        upperB = UpperBound endSlot endSlotClosure
+    pure $ Interval lowerB upperB
+    where
+      convertExtended sysStart eraHist =
+        firstEitherT toError . hoistEither . \case
+          Finite pTime -> do
+            s <- posixTimeToSlot sysStart eraHist pTime
+            pure . Finite . Ledger.Slot . toInteger $ s
+          NegInf -> pure NegInf
+          PosInf -> pure PosInf
+
+      getExtClosure pparams epochInfo sysStart exSlot currentClosure =
+        firstEitherT toError . hoistEither $
+          case exSlot of
+            Finite slot -> do
+              slotsTime <- slotToPOSIXTime pparams epochInfo sysStart (toSlotNo slot)
+              pure $ slotsTime `Ledger.member` ptr
+            NegInf -> pure currentClosure
+            PosInf -> pure currentClosure
+
+-- | Convert `POSIXTime` to `Slot` using `SystemStart` and `EraHistory`.
+posixTimeToSlot ::
+  SystemStart ->
+  EraHistory CardanoMode ->
+  Ledger.POSIXTime ->
+  Either HF.PastHorizonException Ledger.Slot
+posixTimeToSlot sysStart eraHist pTime =
+  -- toRelativeTime checks that pTime >= sysStart via `Control.Exception.assert`
+  let relativeTime = toRelativeTime sysStart (toUtc pTime)
+      (CAPI.EraHistory _ int) = eraHist
+   in HF.interpretQuery int (HF.wallclockToSlot relativeTime)
+        >>= \(s, _, _) -> return (fromSlotNo s)
```

Comment:
Given we're already not point free here from the let, I think this would be cleaner in do notation

Reply:
fixed